docs: streamline A2UI v0.9.1 architecture skill, add AGENTS.md and maintenance skill (#1542)
google/A2UI: .gemini/AGENTS.md

# A2UI Agent Source of Truth (AGENTS.md)

This document is the authoritative guide for all AI agents working within the A2UI repository. It outlines the design philosophy, repository structure, running instructions, and guidelines for keeping this repository and its skills up to date.

---

## 1. What is A2UI?

> **A2UI (Agent-to-User Interface)** is a platform-agnostic, streaming-first UI protocol designed specifically for Large Language Models (LLMs) and autonomous agents.

Reviewer: ```suggestion
**A2UI (Agent-to-User Interface)** is a platform-agnostic, streaming-first UI protocol designed specifically to allow Large Language Models (LLMs) and autonomous agents to generate user interfaces.
```

Unlike traditional APIs that require hardcoded clients or static responses, A2UI enables an agent/server to progressively stream abstract, JSONL-formatted user interface layouts to a client. The client (which can be web-based, desktop, mobile, or a CLI) parses this stream on the fly and renders it dynamically using native components (e.g., Row, Column, Button, Text, Modals, ChoicePickers) and styling rules.

Key capabilities of A2UI include:

- **Streaming UI:** Components and data bind values are painted progressively as they arrive from the server, minimizing user-perceived latency.
- **Two-Way Data Binding:** Seamless synchronization of values between client state and server/agent memory.
- **Function Evaluation (Catalogs):** The protocol supports local validation and execution functions (such as logic gates, formatting, and mathematical operations) registered in Component Catalogs.

---

## 2. Protocol Versioning & Authority

> **IMPORTANT VERSIONING NOTICE:**
>

> > - **v0.9.1** is the **latest published and active protocol version** implemented by our multi-language SDKs, renderers, and sample clients.

Reviewer: Maybe mention 0.9 and 0.8 as well?

Author: Sounds great. Yep let's add that v0.8 is a stable specification, but not supported by the latest version of the SDKs, and that v0.9 is stable, implemented in SDKs, and only very slightly different from v0.9.1.

> - **v0.10** is currently in **draft status** and under active design/development.

> > - **Authority Rule:** Unless explicitly instructed otherwise by the user, **all core architectural decisions, features, refactoring, and code changes must target version v0.9.1 as the absolute authority.** Do not implement changes or conform to schemas outside of the `specification/v0_9_1/` directory unless specifically directed to work on v0.10 or another version.

Reviewer: This seems like it might make it hard to work on other versions: I might say "do this for v0.10" once, and then if I don't keep saying it, it might revert to working on v0.9.1.

I'd soften this a bit so that it defaults to 0.9.1 if the request isn't specific, but doesn't require explicit specification every time.

Author: SGTM, let's soften this to say that you should default to v0.9.1 when working on SDKs unless the user specifies otherwise.

Contains versioned specifications, schemas, evaluation suites, and catalogs:

- **`v0_8/`**: Stable protocol version 0.8.
- **`v0_9/`**: Closed protocol version 0.9.
- **`v0_9_1/`**: Current active protocol version 0.9.1 (contains basic and minimal catalog schemas).
- **`v0_10/`**: Future protocol version under development (draft status).
- Each version subdirectory typically contains:
  - `docs/`: Markdown specification documents (e.g., protocol definitions, renderer guides, styling guides).
  - `json/`: Official JSON Schema definitions for server-to-client messages, client-to-server events, common primitives, and capabilities.
  - `catalogs/`: Component and validation function catalogs (e.g., `basic/catalog.json`, `minimal/catalog.json`).
  - `eval/`: Genkit/TypeScript evaluation suites to test conformance.

### `agent_sdks/` — Server/Agent Integration SDKs

Libraries for agent developers to easily stream A2UI-conforming payloads from their backends:

- **`python/`**: Python SDK (`pyproject.toml` based).
- **`kotlin/`**: Kotlin/JVM SDK (`build.gradle.kts` based).
- **`conformance/`**: Core conformance schema and test suites to validate custom agent library implementations.

> ### `renderers/` — Platform-Specific UI Renderers

Reviewer: We need to do a pass and convert all the mentions of "renderer" to "adapter", to conform with Polina's new glossary.

No need to change it for this PR, just mentioning it as something to keep in mind.

Author: SGTM. Let's stick with renderer for now.

Navigate to any agent sample in `samples/agent/adk/` and launch it using `uv`:

```bash
cd samples/agent/adk/contact_lookup
uv run .
```

_(Ensure you copy `.env.example` to `.env` and configure any required API keys first)._

### Step C: Running a Client (Lit)

With the server running, navigate to a Lit client sample (e.g., `contact` or `shell`) and run the dev server:

```bash
cd samples/client/lit/contact
npm install
npm run dev
```

### Step D: Running a Client (Angular)

To run an Angular sample client, run the following:

```bash
cd samples/client/angular
npm install
npm start -- contact   # Or specify 'restaurant', 'gallery', 'rizzcharts', 'orchestrator'
```

### Step E: Running the Development Tools

- **Visual Editor:** `cd tools/editor && npm install && npm run dev`
- **Visual Composer:** `cd tools/composer && npm install && npm run dev`
- **Payload Inspector:** `cd tools/inspector && npm install && npm run dev`

---

## 5. Maintenance & Update Policy

As the A2UI protocol and codebase evolve, our agent documentation and skills **MUST** be kept perfectly synchronized with the repository's latest state.

### Keep Docs Up To Date

> - Whenever you add, rename, or remove folders, schemas, components, or setup scripts, those modifications **MUST** be updated in `.gemini/AGENTS.md`.

Reviewer: ```suggestion
- Whenever you add, rename, or remove folders, schemas, components, or setup scripts, those modifications **MUST** be updated in `.gemini/AGENTS.md and associated skills files`.
```

Author: Let's do this update.

- After completing any implementation task, always check if `.gemini/AGENTS.md` or any skill under `.gemini/skills/` requires an update, and suggest this to the user.

### Source of Truth Hierarchy

> When verifying protocol, layout, or catalog changes, use the following files under `specification/` as the absolute sources of truth:

Reviewer: ```suggestion
When verifying protocol, layout, or catalog changes to v0.9.1, use the following files under `specification/` as the absolute sources of truth:
```

Otherwise it might get weird when working on other versions.

Author: Great point.

1. **`specification/v0_9_1/docs/a2ui_protocol.md`** — Core semantics and flow.
2. **`specification/v0_9_1/docs/renderer_guide.md`** — Client state layer and subscription lifecycles.

> 3. **`specification/v0_9_1/json/*.json`** — Strict message, capability, and schema envelopes.

Author: Let's move the JSON files to the first position, the absolute source of truth.
